File: JDK11/opentraceability/src/main/java/opentraceability/utility/limitedpoolitem`1.java
package opentraceability.utility;

import java.util.concurrent.TimeUnit;

public class LimitedPoolItem<T> {
    public T value;
    private Action<LimitedPoolItem<T>> disposeAction;
    private long lifetime = TimeUnit.HOURS.toMillis(1);
    private long creationTime = System.currentTimeMillis();

    public LimitedPoolItem(T value, Action<LimitedPoolItem<T>> disposeAction, long lifetime) {
        this.value = value;
        this.disposeAction = disposeAction;
        this.lifetime = lifetime;
    }

    public LimitedPoolItem(T value, Action<LimitedPoolItem<T>> disposeAction) {
        this.value = value;
        this.disposeAction = disposeAction;
    }

    public boolean isExpired() {
        return System.currentTimeMillis() - creationTime > lifetime;
    }

    public void dispose() {
        disposeAction.invoke(this);
    }
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/utility/limitedpoolitem`1.kt
package opentraceability.utility

import java.util.concurrent.TimeUnit

class LimitedPoolItem<T>(
    val value: T,
    private val disposeAction: (LimitedPoolItem<T>) -> Unit,
    private val lifetime: Long = TimeUnit.HOURS.toMillis(1)
) {
    private val creationTime = System.currentTimeMillis()

    val isExpired: Boolean
        get() = System.currentTimeMillis() - creationTime > lifetime

    fun dispose() {
        disposeAction(this)
    }
}
